klippy/kinematics/deltesian.py

```diff
@@ -0,0 +1,184 @@
+        if slow_ratio > 0.:
+            sr2 = slow_ratio ** 2
+            self.slow_x2 = min([math.sqrt((sr2 * a2) / (sr2 + 1))
+                        - axl for a2, axl in zip(arm2, arm_x)]) ** 2
+            self.very_slow_x2 = min([math.sqrt((2 * sr2 * a2) / (2 * sr2 + 1))
+                        - axl for a2, axl in zip(arm2, arm_x)]) ** 2
+            logging.info("Deltesian kinematics: moves slowed past %.2fmm"
+                         " and %.2fmm"
+                         % (math.sqrt(self.slow_x2),
+                            math.sqrt(self.very_slow_x2)))
+        # Setup boundary checks
+        max_velocity, max_accel = toolhead.get_max_velocity()
+        self.max_z_velocity = config.getfloat('max_z_velocity', max_velocity,
+                                              above=0., maxval=max_velocity)
+        self.max_z_accel = config.getfloat('max_z_accel', max_accel,
+                                           above=0., maxval=max_accel)
+        self.axes_min = toolhead.Coord(*[l[0] for l in self.limits], e=0.)
+        self.axes_max = toolhead.Coord(*[l[1] for l in self.limits], e=0.)
+        self.homed_axis = [False] * 3
+        self.set_position([0., 0., 0.], ())
+    def get_steppers(self):
+        return [s for rail in self.rails for s in rail.get_steppers()]
+    def _actuator_to_cartesian(self, sp):
+        arm_x, arm2 = self.arm_x, self.arm2
+        dx, dz = sum(arm_x), sp[1] - sp[0]
+        pivots = math.sqrt(dx**2 + dz**2)
+        # Trilateration w/ reference frame along left to right pivots
+        xt = (pivots**2 + arm2[0] - arm2[1]) / (2 * pivots)
+        zt = math.sqrt(arm2[0] - xt**2)
+        # Rotation and translation of the reference frame
+        x = xt * dx / pivots + zt * dz / pivots - arm_x[0]
+        z = xt * dz / pivots - zt * dx / pivots + sp[0]
+        return [x, z]
+    def _pillars_z_max(self, x):
+        arm_x, arm2 = self.arm_x, self.arm2
+        dz = (math.sqrt(arm2[0] - (arm_x[0] + x)**2),
+              math.sqrt(arm2[1] - (arm_x[1] - x)**2))
+        return min([o - z for o, z in zip(self._abs_endstop, dz)])
+    def calc_position(self, stepper_positions):
+        sp = [stepper_positions[rail.get_name()] for rail in self.rails]
+        x, z = self._actuator_to_cartesian(sp[:2])
+        return [x, sp[2], z]
+    def set_position(self, newpos, homing_axes):
+        for rail in self.rails:
+            rail.set_position(newpos)
+        for n in homing_axes:
+            self.homed_axis[n] = True
+    def home(self, homing_state):
+        homing_axes = homing_state.get_axes()
+        home_xz = 0 in homing_axes or 2 in homing_axes
+        home_y = 1 in homing_axes
+        forceaxes = ([0, 1, 2] if (home_xz and home_y) else
+                     [0, 2] if home_xz else [1] if home_y else [])
+        homing_state.set_axes(forceaxes)
+        homepos = [None] * 4
+        if home_xz:
+            homing_state.set_axes([0, 1, 2] if home_y else [0, 2])
```

Comment:
This is also done above? Perhaps one isn't needed? Not sure if the call above should be split between this and `if home_y` branch.

Reply:
Oops, I forgot to clean up the line 129 after reworking the code. The right one is above the conditions home_xz, home_z. Thanks for catching it!